governance/status.go

```diff
@@ -0,0 +1,413 @@
+			"Workloads: %v,"+
+			"LastUpdated: %v",
+		w.Connectivity, w.Microservices, w.Workloads, w.LastUpdated)
+}
+
+func (w DeviceStatus) IsSame(compare DeviceStatus) bool {
+	if !reflect.DeepEqual(w.Connectivity, compare.Connectivity) {
+		return false
+	} else if !MsArrayIsSame(w.Microservices, compare.Microservices) {
+		return false
+	} else if !WlArrayIsSame(w.Workloads, compare.Workloads) {
+		return false
+	}
+
+	return true
+}
+
+// compare two microservice status arrays
+func MsArrayIsSame(a1 []MicroserviceStatus, a2 []MicroserviceStatus) bool {
+	if len(a1) != len(a2) {
+		return false
+	} else {
+		for _, item1 := range a1 {
+			found := false
+			for _, item2 := range a2 {
+				if item1.IsSame(item2) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// compare two workload array
+func WlArrayIsSame(a1 []WorkloadStatus, a2 []WorkloadStatus) bool {
+	if len(a1) != len(a2) {
+		return false
+	} else {
+		for _, item1 := range a1 {
+			found := false
+			for _, item2 := range a2 {
+				if item1.IsSame(item2) {
+					found = true
+				}
+			}
+			if !found {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// Report the containers status and connectivity status to the exchange.
+func (w *GovernanceWorker) reportStatus() {
+
+	go func() {
+
+		// This is the amount of time for the governance routine to wait.
+		waitTime := w.Config.Edge.ReportDeviceStatusInterval
+		if waitTime == 0 { // never report status when the interval is set to 0
+			glog.V(5).Info("ReportDeviceStatusInterval is 0. The status report to the exchange is turned off.")
+			return
+		}
+
+		for {
+			glog.Info("started the status report to the exchange.")
+
+			var device_status DeviceStatus
+
+			// get connectivity
+			connect := make(map[string]bool, 0)
+			for _, host := range HORIZON_SERVERS {
+				if err := cutil.CheckConnectivity(host); err != nil {
+					glog.Errorf(logString(fmt.Sprintf("Error checking connectivity for %s: %v", host, err)))
+					connect[host] = false
+				} else {
+					connect[host] = true
+				}
+			}
+			device_status.Connectivity = connect
+
+			// get microservice status
+			if ms_status, err := w.getMicroserviceStatus(); err != nil {
+				glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+			} else {
+				device_status.Microservices = ms_status
+			}
+
+			// get workload status
+			if wl_status, err := w.getWorkloadStatus(); err != nil {
+				glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+			} else {
+				device_status.Workloads = wl_status
+			}
+
+			//compare the device status with the one gotten from last time
+			if jbytes, err := json.Marshal(&device_status); err != nil {
+				glog.V(5).Infof(logString(fmt.Sprintf("Failed to convert the device status %v to json: %v", device_status, err)))
+			} else if Device_Status != nil && device_status.IsSame(*Device_Status) {
```

Comment:
You wrote a lot of IsSame and comparison code, which is all very similar and repetitive. Wouldnt it have been easier/better to marshal the status to a byte array and then hash that, and then compare the hash from last time to see if anything has changed? Less duplicate code to maintain, and might even be faster. The reflection APIs (DeepEqual) are known to be slow.

Reply:
The status contains slices and the element order in them may be different. So we cannot use hash to compare them.